### DISCUSSION TITLE
Is this a competitor to Envoy?

### DISCUSSION THREAD

**Discussion by AceHack** [dotnet/yarp#232] [General] [answered]
If so what are you trying to achieve above and beyond what Envoy does?  I'm really interested in this project but just curious.  We use Envoy heavily in our istio service mesh.

**Comment by samsp-msft** [comment, marked as answer]
The differences from Envoy are:
* Its being designed for easy extensibility - both in terms of adding new modules, and also to enable callbacks so that customization can be done in code. For example transforming headers is probably easier to express as a method in C#, rather than having to use regular expressions or other DSL.
* Dynamic routing will be enabled via code extensibility so you can either push routes programatically, or use a callback model and supply the route on a per request basis. Envoy supports the former, but you need to create an gRPC endpoint - in YARP you can do it in-process. 
* It will run equally well on Windows & Linux - This is required for Azure infrastructure which is largely on windows

**Comment by Shahid-Roofi-Khan** [comment]
More differences,
> this is inprocess thing while envoy proxy is out of process proxy
> this is .Net based and subject to GC while Envoy is C++ high performant thing

**Reply by davidfowl** [reply]
> this is inprocess thing while envoy proxy is out of process proxy

Not really. You can host YARP in your ASP.NET core process or in another process. In that sense it's both.

> this is .Net based and subject to GC while Envoy is C++ high performant thing

You should look at the performance numbers of envoy, being C++ doesn't make it faster in all scenarios...
